src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp

```diff
@@ -248,15 +248,18 @@ std::unique_ptr<WriteBufferFromFileBase> S3ObjectStorage::writeObject( /// NOLIN
         std::move(s3_buffer), std::move(finalize_callback), object.absolute_path);
 }
 
-void S3ObjectStorage::findAllFiles(const std::string & path, RelativePathsWithSize & children) const
+void S3ObjectStorage::findAllFiles(const std::string & path, RelativePathsWithSize & children, int max_keys) const
 {
     auto settings_ptr = s3_settings.get();
     auto client_ptr = client.get();
 
     Aws::S3::Model::ListObjectsV2Request request;
     request.SetBucket(bucket);
     request.SetPrefix(path);
-    request.SetMaxKeys(settings_ptr->list_object_keys_size);
+    if (max_keys)
+        request.SetMaxKeys(max_keys);
+    else
+        request.SetMaxKeys(settings_ptr->list_object_keys_size);
```

Comment:
Maybe `request.SetMaxKeys(std::min(max_keys, settings_ptr->list_object_keys_size))` is better, WDYT?

Reply:
Actually no, if `max_keys` is set, then it should not return more keys then was requiested.